### PR TITLE
Support for stdin and getenv in guest

### PR DESCRIPTION
**Pull request by shkoo**
* Actually export the sys_* ABI symbols without mangling them
* Add sys_getenv to get environment variables
* Add with_env_var to ProverOpts to specify environment variables to include for the guest
* Add --env VARNAME=value option to r0vm to specify environment variables to include
* Plumb through stdin() so that std::io::stdin() works in the guest as expected

### CODE REVIEW

File: risc0/zkvm/methods/std/src/bin/standard_lib.rs [hunk 1]
```diff
@@ -12,9 +12,35 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+use std::io::{stdin, stdout, Read, Write};
+
 risc0_zkvm::entry!(main);
 
 pub fn main() {
```
Comment (mothran) [inline risc0/zkvm/methods/std/src/bin/standard_lib.rs:19]: Nice way to integrate std::env testing.

File: risc0/zkvm/platform/src/syscall.rs [hunk 2]
```diff
@@ -393,3 +404,49 @@ pub unsafe extern "C" fn sys_write(fd: u32, write_buf: *const u8, nbytes: usize)
...
     );
 }
+
+/// Retrieves the value of an environment variable, and stores as much
+/// of it as it can it in the memory at [out_words, out_words +
+/// out_nwords).  Returns the length of the value.
+///
+/// This is normally called twice to read an environment variable:
+/// Once to get the length of the value, and once to fill in allocated
+/// memory.
```
Comment (mothran) [inline risc0/zkvm/platform/src/syscall.rs:414]: Heads up, in the future if we integrate clippy, unsafe functions will warn with missing a `# Safety` block in the docstr, if you happen to have a rough idea of the lifetime assumptions of the args, putting them in a Safety block now would be a nice way to jump start that documentation.

reference: https://rust-lang.github.io/rust-clippy/rust-1.56.0/index.html#missing_safety_doc

File: risc0/zkvm/platform/src/syscall.rs [hunk 3]
```diff
@@ -393,3 +404,49 @@ pub unsafe extern "C" fn sys_write(fd: u32, write_buf: *const u8, nbytes: usize)
...
+        a0 as usize
+    }
+}
+
+// Number of words remaining in the heap that haven't yet been allocated.
+static mut HEAP_WORDS_REMAINING: usize = crate::memory::HEAP.len_words();
+
+#[no_mangle]
+pub unsafe extern "C" fn sys_alloc_words(nwords: usize) -> *mut u32 {
+    // SAFETY: Single threaded, so nothing else can touch this while we're working.
```
Comment (mothran) [inline risc0/zkvm/platform/src/syscall.rs:441]: Maybe move to a `# Safety` section in the docstr.

**Review by mothran — APPROVED** [review]
Looks really good.

File: risc0/zkvm/platform/src/syscall.rs (hunk 2 above)
Comment (shkoo) [reply risc0/zkvm/platform/src/syscall.rs:414]: @flaub in https://github.com/risc0/risc0/pull/426#discussion_r1134625722_ indicated a preference for marking all syscalls as "unsafe", even if they don't write to memory or do anything else unsafe.

How do we want to reconcile that?   Should we just add a "# Safety" block saying "This is safe to call!  No worries here."?

Comment (mothran) [reply risc0/zkvm/platform/src/syscall.rs:414]: Hmmm good question, if there is not a straight forward answer now we could just defer until we decide to incorporate clippy.

Comment (flaub) [reply risc0/zkvm/platform/src/syscall.rs:414]: I think we should probably always have a safety comment that goes with any unsafe calls. The justification could be: "There are no known issues with calling this function" or something to that effect.

File: risc0/zkvm/platform/src/syscall.rs (hunk 3 above)
Comment (flaub) [reply risc0/zkvm/platform/src/syscall.rs:441]: I think we also need to get into the habit of docstrings for all public functions

**Pull request merged**